test/integration.test.ts

```diff
@@ -474,3 +484,153 @@ describe('dleq', () => {
+			outputData: createFactory('mintTest')
+		});
+		expectProofsSecretToEqual(proofs, 'mintTest');
+	});
+	test('Manual Factory Send', async () => {
+		function createFactory(pubkey: string): OutputDataFactory {
+			function inner(a: number, k: MintKeys) {
+				return OutputData.createSingleP2PKData({ pubkey }, a, k.id);
+			}
+			return inner;
+		}
+
+		const mint = new CashuMint(mintUrl);
+		const wallet = new CashuWallet(mint);
+
+		const quote = await wallet.createMintQuote(21);
+		await new Promise((res) => setTimeout(res, 1000));
+		const proofs = await wallet.mintProofs(21, quote.quote);
+		const amount = sumProofs(proofs) - wallet.getFeesForProofs(proofs);
+		const { send, keep } = await wallet.send(amount, proofs, {
+			outputData: { send: createFactory('send'), keep: createFactory('keep') }
+		});
+		expectProofsSecretToEqual(send, 'send');
+		expectProofsSecretToEqual(keep, 'keep');
+	});
+	test('Manual BlindingData', async () => {
+		const mint = new CashuMint(mintUrl);
+		const wallet = new CashuWallet(mint);
+		const keys = await wallet.getKeys();
+
+		const quote = await wallet.createMintQuote(40);
+		await new Promise((res) => setTimeout(res, 1000));
+		const proofs = await wallet.mintProofs(40, quote.quote);
+		const data1 = OutputData.createP2PKData({ pubkey: 'key1' }, 10, keys);
+		const data2 = OutputData.createP2PKData({ pubkey: 'key2' }, 10, keys);
+		const { keep, send } = await wallet.send(20, proofs, {
+			outputData: { send: [...data1, ...data2] }
+		});
+		const key1Sends = send.slice(0, data1.length);
+		const key2Sends = send.slice(data1.length);
+		expectProofsSecretToEqual(key1Sends, 'key1');
+		expectProofsSecretToEqual(key2Sends, 'key2');
+	});
+});
+describe('Keep Vector and Reordering', () => {
+	test('Receive', async () => {
+		const mint = new CashuMint(mintUrl);
+		const wallet = new CashuWallet(mint);
+
+		const mintQuote = await wallet.createMintQuote(64);
+		await new Promise((res) => setTimeout(res, 1000));
+		const testOutputAmounts = [8, 4, 8, 2, 8, 2];
+		const testProofs = await wallet.mintProofs(64, mintQuote.quote);
+
+		const { send } = await wallet.send(32, testProofs, { includeFees: true });
+		const receiveProofs = await wallet.receive(
+			{ mint: mintUrl, proofs: send },
+			{ outputAmounts: { keepAmounts: [], sendAmounts: testOutputAmounts } }
+		);
+		receiveProofs.forEach((p, i) => expect(p.amount).toBe(testOutputAmounts[i]));
+	});
+	test('Send', async () => {
+		const mint = new CashuMint(mintUrl);
+		const wallet = new CashuWallet(mint);
+
+		const mintQuote = await wallet.createMintQuote(64);
+		await new Promise((res) => setTimeout(res, 1000));
+		const testOutputAmounts = [8, 4, 8, 2, 8, 2];
+		const testProofs = await wallet.mintProofs(64, mintQuote.quote);
+
+		const fees = wallet.getFeesForProofs(testProofs);
+
+		const { send } = await wallet.send(32, testProofs, {
+			outputAmounts: {
+				sendAmounts: testOutputAmounts,
+				keepAmounts: [16, 8, ...Array(8 - fees).fill(1)]
```

Comment:
can you explain the logic behind this keepAmounts. Is this something the users of cashu-ts will have to do whenever there are fees and they want change?

Reply:
When you use `outputAmounts` you are required to specify amounts that match the keyset and also satisfy the amount constraints exactly. So yes, if your mint takes fees and you want to specify exact output amounts, you need to consider fees.